### PR TITLE
Padding_idx in Embedding supports negative indexing

### LINKED ISSUE #3233: nn.Embedding padding idx doesn't check for negative value
nn.Embedding doesn't check for negative padding index value which might be able to lead to subtle bugs if someone wants to use it with inverted indexing (even though it's not specified that you can).

```
def forward(self, input):
        padding_idx = self.padding_idx
        if padding_idx is None:
            padding_idx = -1
        return self._backend.Embedding.apply(
            input, self.weight,
            padding_idx, self.max_norm, self.norm_type,
            self.scale_grad_by_freq, self.sparse
        )

```

To be clear: defining an embedding layer `emb = nn.Embedding(5,10,padding_idx=-1)` is possible and raises no error. Moreover the last weights are initialized to 0 as you would expect if negative indexing on embedding was working. This can be confusing. 

```
In [44]: emb(Variable(torch.LongTensor([4])))
Out[44]: 
Variable containing:
    0     0     0     0     0     0     0     0     0     0
[torch.FloatTensor of size 1x10]
```

I believe this can be fixed in two ways:

- asserting `padding_idx > 0 or None`
- if `padding_idx < 0` then setting `padding_idx = num_embedding + padding_idx`

### PR DESCRIPTION
**Pull request by dasguptar**
Works for both nn.Module and functional versions, fixes #3233.

### CODE REVIEW

**Comment by colesbury** [comment]
Thanks for the contribution.

I think this would be better implemented in the C++ code. Specifically `embedding`/`embedding_backward` in `aten/src/ATen/native/Embedding.cpp`.

The new functionality also deserves a test. (`test_embedding_padding_idx` in `test_nn.py`)

**Comment by apaszke** [comment]
Pushing it down to the C++ level would be hard, because it's hard to choose the "no padding" value if it's just an `int`. You either need a flag, or choose `INT_MIN` or `INT_MAX`, but these are annoying to get from Python

**Comment by dasguptar** [comment]
@colesbury @apaszke thanks for the feedback. 

I kept the changes in Python because the C++ code uses `-1` to denote no padding, and changing that would require jumping through many more hoops. 

I will try to add a test for the new functionality ASAP.

**Comment by dasguptar** [comment]
Hi @colesbury @apaszke sorry to bother you, but do the changes look fine to you? Can this be merged in that case?

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks @dasguptar!
